sql/core/src/main/scala/org/apache/spark/sql/functions.scala

```diff
@@ -3316,6 +3316,141 @@ object functions {
     ArrayExcept(col1.expr, col2.expr)
   }
 
+  private def expressionFunction(f: Column => Column)
+  : Expression => Expression =
+    x => f(Column(x)).expr
+
+  private def expressionFunction(f: (Column, Column) => Column)
+  : (Expression, Expression) => Expression =
+    (x, y) => f(Column(x), Column(y)).expr
+
+  private def expressionFunction(f: (Column, Column, Column) => Column)
+  : (Expression, Expression, Expression) => Expression =
+    (x, y, z) => f(Column(x), Column(y), Column(z)).expr
+
+  /**
+   * Returns an array of elements after applying a tranformation to each element
+   * in the input array.
+   *
+   * @group collection_funcs
+   */
+  def transform(column: Column, f: Column => Column): Column = withExpr {
+    HigherOrderUtils.transform(column.expr, expressionFunction(f))
+  }
+
+  /**
+   * Returns an array of elements after applying a tranformation to each element
+   * in the input array.
+   *
+   * @group collection_funcs
+   */
+  def transform(column: Column, f: (Column, Column) => Column): Column = withExpr {
+    HigherOrderUtils.transform(column.expr, expressionFunction(f))
+  }
+
+  /**
+   * Returns whether a predicate holds for one or more elements in the array.
+   *
+   * @group collection_funcs
+   */
+  def exists(column: Column, f: Column => Column): Column = withExpr {
```

Comment:
But how do we support this in Java?

Reply:
Could we change the signatures to accept `scala.runtime.AbstractFunction`s instead to avoid using the Function traits?

Reply:
Let's add `(Scala-specific) ` at least for each doc. BTW, please take a look for style guide at https://github.com/databricks/scala-style-guide

Reply:
Actually a better idea would probably be to use java functional interfaces.

```java
@FuncitonalInterface
interface Function3[T1, T2, T3, R] {
  R apply(T1 t1, T2 t2, T3 t3);
}

Column exists(Column column, Function3[Column, Column, Column, Column] f) = ...
```

Reply:
And of course we would use the existing functional interfaces first from `java.util.function`, but I don't think there are any that accept three parameters likes some of the functions here require.

Reply:
It appears these interfaces already exist in the source tree: https://github.com/apache/spark/blob/v2.4.0/core/src/main/java/org/apache/spark/api/java/function/Function3.java

I'll come back later to add java-specific apis that utilizes these.